### ISSUE TITLE
Confirmation not working

### ISSUE THREAD

**Issue by rpolenthon** [arachnetech/homebridge-mqttthing#214]
I set up the confirmation like this:
```
{
  "accessory" : "mqttthing",
  "confirmationPeriodms" : 1000,
  "mqttPubOptions" : {
    "retain" : true
  },
  "name" : "WZ Licht Vorne",
  "offValue" : "off",
  "onlineValue" : "true",
  "onValue" : "on",
  "retryLimit" : 3,
  "topics" : {
    "getOn" : "shellies/shelly1-2C6BF0/relay/0",
    "getOnline" : "shellies/shelly1-2C6BF0/online",
    "setOn" : "shellies/shelly1-2C6BF0/relay/0/command"
  },
  "type" : "switch",
  "url" : "http://192.168.178.113:1883"
}
```

The retry attempts are working perfectly but sadly it doesn`t go into the offline state if no answer gets received.

Did I configure something wrong?

Note: To test the confirmation, I changed the topic to a not existing one, so it is like the regular device would not be connected to the broker in case of a failure.

**Comment by arachnetech** [comment]
The Home app only checks the online state when launched - so you’ll need to close and re-open the Home app to see this.

**Comment by rpolenthon** [comment]
Ok, I will try that again someday, but I thought that I already tried that.

For now I will close the issue, thanks.

**Comment by th3cube** [comment]
For me its also not working. Even when the device has no Power the State isn´t reflected in the Home App

**Comment by arachnetech** [comment]
It relies on your device setting an appropriate LWT message to be published by the MQTT broker to notify mqttthing that it’s offline, and your device publishing an online message when it is available. 

Also the Home App doesn’t notice online/offline state immediately - you may need to restart it.

**Comment by rpolenthon** [comment]
Maybe there was a misunderstanding. I am not talking about the online topic confirmation, this is working well.

I am/was talking about this confirmation feature: https://github.com/arachnetech/homebridge-mqttthing#confirmation

**Comment by arachnetech** [comment]
You wrote:
> The retry attempts are working perfectly but sadly it doesn`t go into the offline state if no answer gets received.

Is the online/offline state after the retries not the problem? If so, what is?

**Comment by rpolenthon** [comment]
Online / Offline detection is split into two parts. One part is the online topic which gets the LWT from the broker. This is working fine.

The second part is the "offline" detection based on the specified confirmation retries incl. confirmation period. This isn't working. For my understanding, the plugin should set the accessory to the offline state in case the getTopic does not deliver the same value as the setTopic. But this isn't working. It is not the typical Home App issue. I restarted the Home app multiple times and I also checked it again after several hours.

Maybe I just understand the new confirmation feature wrong...

**Comment by arachnetech** [comment]
Sorry, meant to add: you’re right that the other online/offline mechanism using LWT may be irrelevant here, but I’m not quite sure what isn’t working for you.

**Comment by arachnetech** [comment]
You’re right: I think the confirmation feature does do that, although I don’t seem to mention it in the readme. Checking...

**Comment by arachnetech** [comment]
Just confirmed with the 'Test outlet' in my test config:

```
        {
            "accessory": "mqttthing",
            "type": "outlet",
            "name": "Test outlet",
            "url": "http://192.168.10.35:1883",
            "mqttPubOptions": {
                "retain": true
            },
            "topics": {
                "getOn": "test/outlet/getOn",
                "setOn": "test/outlet/setOn",
                "getInUse": "test/outlet/getInUse",
                "getWatts": "test/outlet/getWatts",
                "getVolts": "test/outlet/getVolts",
                "getAmperes": "test/outlet/getAmperes"
            },
            "integerValue": true,
            "history": true,
            "confirmationPeriodms": 2000,
            "retryLimit": 10
        },
```

... and it is setting the accessory to show as offline automatically. *It is necessary to close and re-open the Home app to see this.*

![image](https://user-images.githubusercontent.com/24389384/79054251-605f6280-7c3b-11ea-89f7-8df166d05f00.png)

I'm not sure how my config differs to yours.

**Comment by arachnetech** [comment]
Ooh, with your switch it's not working for me either!

`[4/11/2020, 9:32:26 PM] [WZ Licht Vorne] Unresponsive - no confirmation message received on shellies/shelly1-2C6BF0/relay/0. Expecting [on].`

... but doesn't go to 'No Response'...

OK, I see your problem!

**Comment by arachnetech** [comment]
It's working _as intended_ - or at least as I intended when I wrote it (but then forgot).

The confirmation offline indication only happens if you don't have a `getOnline` topic. The logic is that if you have a `getOnline` topic you'll use this to indicate when the accessory is offline - so it doesn't need to try to work it out from lack of confirmation messages.

If you remove your `"getOnline" : "shellies/shelly1-2C6BF0/online",` line then it goes offline as you'd expect (after reopening the Home app):

![image](https://user-images.githubusercontent.com/24389384/79054669-b2ee4e00-7c3e-11ea-8fa5-b82964227f9f.png)

**Comment by arachnetech** [comment]
I've added a new setting (`confirmationIndicateOffline`) allowing this behaviour to be configured. If it's not specified, mqttthing will work as previously - i.e. it will only show the accessory as unresponsive when no confirmation is received if there's no `getOnline` topic. Setting `confirmationIndicateOffline` to `true` will make it behave as you expected.

**Comment by rpolenthon** [comment]
Perfect!
Thanks a lot. Now I know what I will do this evening... updating over 40 Shelly configs.

BTW, as I have several different types of Shellies working here, I will soon make a pull request for a wiki update ;) Some of the configurations may be complicated for some users, as they need a bit more complex apply functions.

**Comment by randy3k** [comment]
I have a related but not identical issue
```
[10/30/2020, 00:32:33] [Apple TV] Publishing MQTT: appletv/cmnd/toggle = false
[10/30/2020, 00:32:36] [Apple TV] Setting accessory state to offline
[10/30/2020, 00:32:36] [Apple TV] Publishing MQTT: appletv/cmnd/toggle = false
[10/30/2020, 00:32:36] [Apple TV] Received MQTT: appletv/stat = false
[10/30/2020, 00:32:36] [Apple TV] Received MQTT: appletv/stat = false
[10/30/2020, 00:32:39] [Apple TV] Publishing MQTT: appletv/cmnd/toggle = false
[10/30/2020, 00:32:39] [Apple TV] Received MQTT: appletv/stat = false
[10/30/2020, 00:32:42] [Apple TV] Publishing MQTT: appletv/cmnd/toggle = false
[10/30/2020, 00:32:42] [Apple TV] Received MQTT: appletv/stat = false
[10/30/2020, 00:32:45] [Apple TV] Unresponsive - no confirmation message received on appletv/stat. Expecting [false].
```

My device has sent back the expected topic but `mqttthing` still thinks that I haven't sent the confirmation. Not sure why.

**Comment by randy3k** [comment]
I have a related but not identical issue
```
[10/30/2020, 00:32:33] [Apple TV] Publishing MQTT: appletv/cmnd/toggle = false
[10/30/2020, 00:32:36] [Apple TV] Setting accessory state to offline
[10/30/2020, 00:32:36] [Apple TV] Publishing MQTT: appletv/cmnd/toggle = false
[10/30/2020, 00:32:36] [Apple TV] Received MQTT: appletv/stat = false
[10/30/2020, 00:32:36] [Apple TV] Received MQTT: appletv/stat = false
[10/30/2020, 00:32:39] [Apple TV] Publishing MQTT: appletv/cmnd/toggle = false
[10/30/2020, 00:32:39] [Apple TV] Received MQTT: appletv/stat = false
[10/30/2020, 00:32:42] [Apple TV] Publishing MQTT: appletv/cmnd/toggle = false
[10/30/2020, 00:32:42] [Apple TV] Received MQTT: appletv/stat = false
[10/30/2020, 00:32:45] [Apple TV] Unresponsive - no confirmation message received on appletv/stat. Expecting [false].
```

My device has sent back the expected topic but `mqttthing` still thinks that I haven't sent the confirmation. Not sure why.

```
        {
            "type": "switch",
            "name": "Apple TV",
            "username": "xxxx",
            "password": "xxxx",
            "logMqtt": true,
            "topics": {
                "getOn": "appletv/stat",
                "setOn": "appletv/cmnd/toggle"
            },
            "confirmationPeriodms": 3000,
            "accessory": "mqttthing"
        }
```

**Comment by rpolenthon** [comment]
Maybe this working example helps you:

```
{
  "accessory" : "mqttthing",
  "confirmationIndicateOffline" : true,
  "confirmationPeriodms" : 1000,
  "name" : "Klingeln",
  "offValue" : "off",
  "onlineValue" : "true",
  "onValue" : "on",
  "retryLimit" : 2,
  "topics" : {
    "getOn" : "shellies/shelly1-259E4C/relay/0",
    "getOnline" : "shellies/shelly1-259E4C/online",
    "setOn" : "shellies/shelly1-259E4C/relay/0/command"
  },
  "type" : "switch",
  "url" : "http://192.168.86.15:1883"
}
```

**Comment by randy3k** [comment]
I didn't setup the `getOnline` topic, but it doesn't seem to be required by judging from the documentation.

**Comment by rpolenthon** [comment]
But you may need `retryLimit`

**Comment by arachnetech** [comment]
There was an issue with mqttthing where it failed to accept true/false confirmation values. This was fixed in 1.1.21.
